allow unsetting options in PlotDataItem and PlotCurveItem

NilsNemitz opened the pull request:
This is a quick attempt to address #2039: Various plotting parameters can be set, but cannot be unset, especially when going through the `PlotDataItem` interfaces.

The origin of the problem is that much of the code implicitly assumes that a new element is initialized from zero, and there are explicit checks that cancel out of the update early when parameters are `None`.

This problem is present in both `PlotDataItem` and `PlotCurveItem`, but the faulty shortcut checks are particularly used in the `setData()` methods. I think the problem is worse when going through a `PlotDataItem`, because this tends to call `PlotCurveItem.setData()` with any changes, 

I made a quick pass over the code and removed problematic shortcut checks for `None`, as well as unconditional `mkBrush` / `mkPen` calls that prevented setting these to explicit None values.

The tests seem to pass, and this *seems* to address the test cases described in #2039.
But I have probably missed a few things, so some additional testing and feedback would be appreciated!

@pijyoi , do you immediately see anything that this misses?


Closes #2039 ... once it works.

j9ac9k commented:
@NilsNemitz do you think `skipFiniteCheck` should be a sticky option along w/ the others?

NilsNemitz commented:
Hmm. So far, I always read "skipFiniteCheck" as a hint for just the current dataset.

But there is an argument to be made for turning it sticky:
In many cases, your data source will never produce non-finite values. Being forced to set the hint for each `setData()` call is needless boilerplate then. 
Doing that is only useful if 
- non-finite values occur only occasionally
- the user code already knows when they are present. And where that rare case applies, a sticky flag can still be updated with every call.
So I think a sticky flag might be better.

NilsNemitz commented on pyqtgraph/graphicsItems/PlotCurveItem.py:
```diff
@@ -514,18 +536,18 @@ def updateData(self, *args, **kargs):
-
-        self.opts['skipFiniteCheck'] = kargs.get('skipFiniteCheck', False)
+        if 'skipFiniteCheck' in kargs:
+            self.opts['skipFiniteCheck'] = kargs['skipFiniteCheck']
```
This updates skipFiniteCheck only on request. A default False value will be used if it is never explicitly set.

NilsNemitz commented:
The latest commit should make `skipFiniteCheck` a sticky flag that can be set once, and will still be effective during later setData() calls. It also adds a `setSkipFiniteTest()` method (for both `PlotDataItem` and `PlotCurveItem`) to allow setting/unsetting this outside of `setData()`.

I have updated the documentation for the logic and effect, and also added some text that explains the recently added handling of wide lines.

I would also suggest adding a final one-year expiration period for the long-deprecated options `stepMode=True`, `decimate` and `identical`. I have added that for now. The main motivation would be to keep the code simple. If there is a reason to keep these mostly non-functional parameters, please let me know!

pijyoi commented on pyqtgraph/graphicsItems/PlotDataItem.py:
```diff
@@ -259,18 +261,37 @@ def __init__(self, *args, **kargs):
+        require merging the entire plot into a single entity before the alpha value can be applied. For plots with more
+        than a few hundred points, this can result in excessive slowdown.
+
+        Since version 0.12.3, this slowdown is automatically avoided by an algorithm that draws line segments
```
Since 0.12.4 (or whichever version comes after 0.12.3)

j9ac9k commented on pyqtgraph/graphicsItems/PlotCurveItem.py:
```diff
@@ -496,7 +515,10 @@ def updateData(self, *args, **kargs):
         if self.opts['stepMode'] in ("center", True):  ## check against True for backwards compatibility
             if self.opts['stepMode'] is True:
                 import warnings
```
I know you didn't just add this, but would you mind removing this import here given the static code checker is complaining about it?

pijyoi commented on pyqtgraph/graphicsItems/PlotCurveItem.py:
```diff
@@ -416,53 +421,56 @@ def setSkipFiniteCheck(self, skipFiniteCheck):
+                        by line segments. 
+                        
+                            | 'all' (default) indicates full connection. 
+                            | 'pairs' omits even-numbered segments. 
```
The previous docs say that even-numbered segments are drawn, presumably counting from zero.

NilsNemitz replied:
Hmm. If you naturally start counting from zero, then neither of these descriptions seems particularly helpful.

"  'pairs' draws one separate line segment for each two points given, starting with the first point.   " ?

pijyoi replied:
I suppose ``` 'pairs' draws one separate line segment for each two points given.``` would be good enough. Don't think there is any doubt about it starting from the first point.

NilsNemitz replied:
Agreed, and changed.

j9ac9k commented:
Thanks @NilsNemitz you clearly have been busy ...love the documentation updates too 👍🏻  merging.
